File: old_samples/sample-core/src/main/java/com/quickblox/sample/core/utils/SystemPermissionHelper.java
package com.quickblox.sample.core.utils;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;

import java.util.ArrayList;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

public class SystemPermissionHelper {
    public static final int PERMISSIONS_FOR_SAVE_FILE_IMAGE_REQUEST = 1;
    public static final int PERMISSIONS_FOR_TAKE_PHOTO_REQUEST = 2;
    public static final int PERMISSIONS_FOR_CALL_REQUEST = 3;

    private Activity activity;
    private Fragment fragment;

    public SystemPermissionHelper(Activity activity) {
        this.activity = activity;
    }

    public SystemPermissionHelper(Fragment fragment) {
        this.fragment = fragment;
    }

    public boolean isSaveImagePermissionGranted() {
        return isPermissionGranted(Manifest.permission.WRITE_EXTERNAL_STORAGE);
    }

    public boolean isCameraPermissionGranted() {
        return isPermissionGranted(Manifest.permission.CAMERA);
    }

    public boolean isCallPermissionsGranted() {
        return isPermissionGranted(Manifest.permission.RECORD_AUDIO) && isPermissionGranted(Manifest.permission.CAMERA);
    }

    private boolean isPermissionGranted(String permission) {
        if (fragment != null) {
            return ContextCompat.checkSelfPermission(fragment.getContext(), permission) == PackageManager.PERMISSION_GRANTED;
        } else {
            return ContextCompat.checkSelfPermission(activity.getApplicationContext(), permission) == PackageManager.PERMISSION_GRANTED;
        }
    }

    public void requestPermissionsForCallByType() {
        checkAndRequestPermissions(PERMISSIONS_FOR_CALL_REQUEST, Manifest.permission.RECORD_AUDIO, Manifest.permission.CAMERA);
    }

    public void requestPermissionsForSaveFileImage() {
        checkAndRequestPermissions(PERMISSIONS_FOR_SAVE_FILE_IMAGE_REQUEST, Manifest.permission.WRITE_EXTERNAL_STORAGE);
    }

    public void requestPermissionsTakePhoto() {
        checkAndRequestPermissions(PERMISSIONS_FOR_TAKE_PHOTO_REQUEST, Manifest.permission.CAMERA);
    }

    private void checkAndRequestPermissions(int requestCode, String... permissions) {
        if (collectDeniedPermissions(permissions).length > 0) {
            requestPermissions(requestCode, collectDeniedPermissions(permissions));
        }
    }

    private String[] collectDeniedPermissions(String... permissions) {
        ArrayList<String> deniedPermissionsList = new ArrayList<>();
        for (String permission : permissions) {
            if (!isPermissionGranted(permission)) {
                deniedPermissionsList.add(permission);
            }
        }

        return deniedPermissionsList.toArray(new String[deniedPermissionsList.size()]);
    }

    private void requestPermissions(int requestCode, String... permissions) {
        if (fragment != null) {
            fragment.requestPermissions(permissions, requestCode);
        } else {
            ActivityCompat.requestPermissions(activity, permissions, requestCode);
        }
    }
}

File: sample-videochat-kotlin/src/main/java/com/quickblox/sample/videochatkotlin/utils/SystemPermissionHelper.kt
package com.quickblox.sample.videochatkotlin.utils

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.util.Log
import java.util.*


class SystemPermissionHelper(val activity: Activity) {
    val TAG = SystemPermissionHelper::class.java.simpleName

    companion object {
        const val PERMISSIONS_FOR_CALL_REQUEST = 15
    }

    fun requestPermissionsForCallByType() {
        checkAndRequestPermissions(PERMISSIONS_FOR_CALL_REQUEST, Manifest.permission.RECORD_AUDIO, Manifest.permission.CAMERA)
    }

    fun checkAndRequestPermissions(requestCode: Int, vararg permissions: String) {
        if (collectDeniedPermissions(*permissions).size > 0) {
            requestPermissions(requestCode, *collectDeniedPermissions(*permissions))
        }
    }

    private fun collectDeniedPermissions(vararg permissions: String): Array<String> {
        val deniedPermissionsList = ArrayList<String>()
        for (permission in permissions) {
            if (!isPermissionGranted(permission)) {
                deniedPermissionsList.add(permission)
            }
        }

        return deniedPermissionsList.toTypedArray()
    }

    fun requestPermissions(requestCode: Int, vararg permissions: String) {
        Log.v(TAG, "request Permissions for activity " + activity.javaClass.simpleName)
        ActivityCompat.requestPermissions(activity, permissions, requestCode)
    }

    fun isAllCameraPermissionGranted(): Boolean {
        return isPermissionGranted(Manifest.permission.CAMERA)
    }

    fun isPermissionGranted(permission: String): Boolean {
        return ContextCompat.checkSelfPermission(activity.getApplicationContext(), permission) == PackageManager.PERMISSION_GRANTED
    }

    fun collectDeniedPermissionsFomResult(permissions: Array<String>, grantResults: IntArray): ArrayList<String> {
        val deniedPermissions = ArrayList<String>()
        for (i in permissions.indices) {
            if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                deniedPermissions.add(permissions[i])
            }
        }

        return deniedPermissions
    }
}
